src/roompanel/RoomManager.cpp

```diff
@@ -0,0 +1,252 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright (C) 2021 The MMapper Authors
+// Author: Massimiliano Ghilardi <massimiliano.ghilardi@gmail.com> (Cosmos)
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <QtCore>
+
+#include "../global/StringView.h"
+#include "../proxy/GmcpMessage.h"
+#include "RoomManager.h"
+#include "RoomMobs.h"
+
+RoomManager::RoomManager(QObject *const /*parent*/)
+    : QObject{nullptr}
+    , m_room{std::make_unique<RoomMobs>(nullptr)}
+{}
+
+RoomManager::~RoomManager() = default;
+
+void RoomManager::slot_reset()
+{
+    m_room->resetMobs();
+}
+
+void RoomManager::updateWidget()
+{
+    emit sig_updateWidget();
+}
+
+void RoomManager::slot_parseGmcpInput(const GmcpMessage &msg)
+{
+    switch (msg.getType()) {
+    case GmcpMessageTypeEnum::ROOM_CHARS_ADD:
+        parseGmcpAdd(msg);
+        break;
+    case GmcpMessageTypeEnum::ROOM_CHARS_REMOVE:
+        parseGmcpRemove(msg);
+        break;
+    case GmcpMessageTypeEnum::ROOM_CHARS_SET:
+        parseGmcpSet(msg);
+        break;
+    case GmcpMessageTypeEnum::ROOM_CHARS_UPDATE:
+        parseGmcpUpdate(msg);
+        break;
+    default:
+        break;
+    }
+}
+
+void RoomManager::parseGmcpAdd(const GmcpMessage &msg)
+{
+    showGmcp(msg);
+    if (isInvalidJson(msg)) {
+        return;
+    }
+    const QJsonDocument &doc = msg.getDocument();
+    if (!doc.isObject()) {
+        if (false) {
+            qWarning().noquote() << "RoomManager received GMCP" << msg.getName().toQString()
+                                 << "containing invalid Json: expecting object, got"
+                                 << msg.getJson().toQString();
+        }
+        return;
+    }
+    addMob(doc.object());
+}
+
+void RoomManager::parseGmcpRemove(const GmcpMessage &msg)
+{
+    showGmcp(msg);
+    // payload is a single number (usually followed by a space), not an array or object
+    // thus parsed msg.getDocument() is not valid: use msg.getJson() instead
+    const std::string &str = msg.getJson().getStdString();
+    const unsigned long num = std::stoul(str);
```

Comment:
Would doing some bound checks here help anything?

Reply:
Yes, we can ignore strings that contain trail junk or that parse to out-of-range integers. I'll add that.

Reply:
Added.